### DISCUSSION TITLE
Algolia search is broken

### DISCUSSION THREAD

**Discussion by baumannzone** [vuejs/vitepress#2030] [Q&A] [answered]
### Describe the bug

I am experiencing an issue with the integrated search component (Algolia).
The search function is not working properly, as it returns **zero results** when I use the web interface in my recently created project with Vitepress. 

However, if I perform a search using the API or on Algolia's website, the search function works correctly and returns results.

It's possible that there may be an issue with the search component in Vitepress that's causing this problem. 

I wanted to bring this to your attention in the hopes that it can be addressed and fixed as soon as possible. 

### Reproduction

- Create an Algolia account
- Configure the algolia section with index, API key and app id.
  - ```
      appId: '5D6FRGxxxx',  
      apiKey: 'e04451b589c484090xxxxxxxxxxxxxxx',  
      indexName: 'csspills'
      ```
- Create the index (both manually / via API)
  - ```
    [
      {
        title: 'Pill 1 CLI',
        description: 'Create a new project with the CLI',
        tags: [ 'demo1', 'cli' ]
      },
      {
        title: 'Pill 2 AWS',
        description: 'Create a new AWS project from the command line',
        tags: [ 'demo2', 'AWS' ]
      }
    ]
    ```
- Start the local server
- Search 

![image](https://user-images.githubusercontent.com/5422102/220378526-6c84ae52-ccab-4270-b25f-73f8135e60b5.png)

### Expected behavior

Show results after searching

### System Info

```shell
System:
    OS: macOS 13.2.1
    CPU: (8) x64 Apple M1
    Memory: 7.73 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.13.2 - ~/.nvm/versions/node/v16.13.2/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 8.1.2 - ~/.nvm/versions/node/v16.13.2/bin/npm
  Browsers:
    Brave Browser: 107.1.45.133
    Chrome: 109.0.5414.119
    Chrome Canary: 112.0.5609.0
    Firefox: 95.0.2
    Firefox Developer Edition: 108.0
    Safari: 16.3
```

### Additional context

UI Request: 
```
https://5d6frgc6t9-dsn.algolia.net/1/indexes/*/queries?x-algolia-agent=Algolia%20for%20JavaScript%20(4.8.5)%3B%20Browser%20(lite)%3B%20docsearch%20(3.3.3)%3B%20docsearch-react%20(3.3.3)%3B%20docsearch.js%20(3.3.3)&x-algolia-api-key=e04451b589c484090xxxxxxxxxxxxxxx&x-algolia-application-id=5D6FRGxxxx
```
Request Method: `POST`
Status Code: `200 OK`

![image](https://user-images.githubusercontent.com/5422102/220379976-4dda3eff-b8ab-4b63-98b0-6282d58bcef5.png)

API Request (Node.js)
![image](https://user-images.githubusercontent.com/5422102/220382548-00a172bf-5c67-4b09-9d55-3b37bd27123c.png)

```
{
  hits: [
    {
      title: 'Pill 2 AWS',
      description: 'Create a new AWS project from the command line',
      tags: [Array],
      objectID: '85344354000',
      _highlightResult: [Object]
    }
  ],
  nbHits: 1,
  page: 0,
  nbPages: 1,
  hitsPerPage: 20,
  exhaustiveNbHits: true,
  exhaustiveTypo: true,
  exhaustive: { nbHits: true, typo: true },
  query: 'AWS',
  params: 'query=AWS',
  renderingContent: {},
  processingTimeMS: 1,
  processingTimingsMS: {
    getIdx: { load: [Object], total: 1 },
    request: { roundTrip: 30 },
    total: 1
  },
  serverTimeMS: 1
}
```

### Validations

- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.vuejs.org).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

**Comment by baumannzone** [comment]
Please let me know if there is any further information or details I can provide to assist with troubleshooting this issue."

**Comment by brc-dd** [comment]
your index needs the `lang` information too as we are filtering based on that

**Comment by baumannzone** [comment]
Hi, @brc-dd! Thanks for your reply.
Where should I put that `lang` key and what values does it accept, any link to the doc?

My question is, why does it work through the API?

**Comment by brc-dd** [comment]
I'm not sure how you're creating that index, but if you'd run algolia's crawler it would've indexed the html lang attribute too. 

It's working through the API because you haven't specified `facetFilters: ['lang:en-US']` in your request.

**Comment by baumannzone** [comment]
Im using the free quote, I don't have access to the crawler. I'm completely new to Algolia, I may be missing something (?)

**Comment by brc-dd** [comment]
Just a clarification on our side: we only support https://docsearch.algolia.com/  (docsearch is free for open-source libraries). If that's the case for you, then you can use https://crawler.algolia.com/admin/ to access crawler configuration.

If that's not the case, then you will need to build your own search component or use any of the ones mentioned at #670

**Comment by baumannzone** [comment]
Thanks for your clarification. It's a public repo and it's open source. 
Here is [the website](https://onebeyond.github.io/css-pills/) and this is [the repo](https://github.com/onebeyond/css-pills).

So if I'm not wrong, I should have access to the algolia crawler, right?

**Comment by brc-dd** [comment]
Yeah, you should have, can you try logging in that admin dashboard?

**Comment by baumannzone** [comment]
Thank you. Actually, I dont have access. But it looks like an auth error. I will try later.

**Comment by brc-dd** [comment]
If it's a blocker than try downgrading to `vitepress@1.0.0-alpha.36` or older. Those versions should work. Or you can also [patch](https://github.com/ds300/patch-package) vitepress to delete this line:

https://github.com/vuejs/vitepress/blob/e8e87215aceb80c8275985aeb40062641d30b626/src/client/theme-default/components/VPAlgoliaSearchBox.vue#L39

**Comment by baumannzone** [comment]
I just requested access to the crawler. It seems to be that was the problem.
If that doesn't solve my problem, I'll be creating a custom search component. Let's keep this issue open until I figure out what's going on.

As feedback, I think Vitepress docs should provide more context about how Algolia should work.

**Comment by MarkusKeck** [comment]
If you really need a fast solution and you need the newest vitepress version you can try `vitepress-plugin-search`.
It is an offline search so you dont need to register anything, in the meantime you can try to configure angolia.

```ts
//vite.config.ts
import { SearchPlugin } from "vitepress-plugin-search";
import { defineConfig } from "vite";

//default options
var options = {
  ...flexSearchIndexOptions,
  previewLength: 62,
  buttonLabel: "Search",
  placeholder: "Search docs",
};

export default defineConfig({
  plugins: [SearchPlugin(options)],
});
```

**Reply by Richard-Zhang1019** [reply]
guys, I use your solution to resolve my search question, but when toggle  docs language it can't work perfect

**Reply by brc-dd** [reply]
@Richard-Zhang1019 just wait for a week or so, we'll ship local search out-of-box!

**Reply by Richard-Zhang1019** [reply]
Amazing information！！

**Reply by Richard-Zhang1019** [reply]
I am looking forward to this feature！！！

**Comment by brc-dd** [comment]
> As feedback, I think Vitepress docs should provide more context about how Algolia should work.

Yeah, the docs are missing many things. I agree that it's sort of ironic considering we are building a docs-centered framework 😅. But yeah we will update that with an official crawler config and hopefully support for local search as well.

After you get access, try adding attributesForFaceting (ref. https://github.com/vuejs/vitepress/issues/1935#issuecomment-1433765099)

**Comment by baumannzone** [comment]
Thank you very mucho, amigo. I'll let you know how it goes. 

@MarkusKeck I want to try using the default search component, I'll try your solution if I can't get it to work.

**Comment by baumannzone** [comment, marked as answer]
I just got a message, it says my site won't be accepted bcs is under construction and has tons of `lorem ipsum`s text. But I think when I finish creating the website, I will be given access to the tool.

**Reply by baumannzone** [reply]
It works: https://onebeyond.github.io/css-pills/ 

https://github.com/vuejs/vitepress/discussions/2030#discussioncomment-5329743

**Comment by brc-dd** [comment]
Let's move this to discussion. If there is any issue you need some help you can directly post in our discord too. For now you can either wait for approval or use the plugin Markus mentioned above.

**Comment by lslzl3000** [comment]
@brc-dd I tried https://docsearch.algolia.com/docs/legacy/run-your-own/#integration  with docker

the crawl config:
```json
{
  "index_name": "orillusion",
  "start_urls": [
    {
      "url": "https://xxxx.com/guide/",
      "variables": {
        "lang": [
          "zh"
        ],
        "version": [
          "latest"
        ]
      },
      "page_rank": 5,
      "selectors_key": "doc"
    },
    {
      "url": "https://xxxx.com/api/",
      "variables": {
        "lang": [
          "en",
          "zh"
        ],
        "version": [
          "latest"
        ]
      },
      "page_rank": 4,
      "selectors_key": "api"
    }
  ],
  "selectors": {
    "doc":{
      "lvl0": ".vp-doc h1",
      "lvl1": ".vp-doc h2",
      "lvl2": ".vp-doc h3",
      "lvl3": ".vp-doc h4",
      "text": ".vp-doc p, .vp-doc ul, .vp-doc ol"
    },
    "api":{
      "lvl0": ".vp-doc h1",
      "lvl1": ".vp-doc h3"
    }
  }
}
```
The record in Algolia dashboard works fine
and the config in vitepress (1.0.0-alpha.60):
```js
algolia: {
      appId: 'xxxxxx',
      apiKey: 'xxxxxxx',
      indexName: 'orillusion',
      searchParameters:{
        facetFilters: ["lang:zh", "version:latest"]
      }
}
```

But I can't get any search record in vitepress, do I miss someting?

**Reply by brc-dd** [reply]
I'm not completely sure if the legacy scrapper is compatible with DocSearch v3, but can you try running `./docsearch playground` to see if the generated index is working? Also when you're searching from VitePress are you getting any errors on the browser console?

**Comment by lslzl3000** [comment]
`./docsearch playground` ?  in where?
No errors or logs on browser console, just no hits`{"results":[{"hits":[],"nbHits":0,"page":0,"nbPages":0, ...`

**Reply by brc-dd** [reply]
Can you try if https://github.com/vuejs/vitepress/issues/1935#issuecomment-1433765099 works?

**Reply by brc-dd** [reply]
FWIW, VitePress' crawler config looks like this:

```ts
new Crawler({
  appId: "...",
  apiKey: "...",
  rateLimit: 8,
  startUrls: ["https://vitepress.dev/"],
  renderJavaScript: false,
  sitemaps: [],
  exclusionPatterns: [],
  ignoreCanonicalTo: false,
  discoveryPatterns: ["https://vitepress.dev/**"],
  schedule: "at 05:10 on Saturday",
  actions: [
    {
      indexName: "vitepress",
      pathsToMatch: ["https://vitepress.dev/**"],
      recordExtractor: ({ $, helpers }) => {
        return helpers.docsearch({
          recordProps: {
            lvl1: ".content h1",
            content: ".content p, .content li",
            lvl0: {
              selectors: "",
              defaultValue: "Documentation",
            },
            lvl2: ".content h2",
            lvl3: ".content h3",
            lvl4: ".content h4",
            lvl5: ".content h5",
          },
          indexHeadings: true,
        });
      },
    },
  ],
  initialIndexSettings: {
    vitepress: {
      attributesForFaceting: ["type", "lang"],
      attributesToRetrieve: ["hierarchy", "content", "anchor", "url"],
      attributesToHighlight: ["hierarchy", "hierarchy_camel", "content"],
      attributesToSnippet: ["content:10"],
      camelCaseAttributes: ["hierarchy", "hierarchy_radio", "content"],
      searchableAttributes: [
        "unordered(hierarchy_radio_camel.lvl0)",
        "unordered(hierarchy_radio.lvl0)",
        "unordered(hierarchy_radio_camel.lvl1)",
        "unordered(hierarchy_radio.lvl1)",
        "unordered(hierarchy_radio_camel.lvl2)",
        "unordered(hierarchy_radio.lvl2)",
        "unordered(hierarchy_radio_camel.lvl3)",
        "unordered(hierarchy_radio.lvl3)",
        "unordered(hierarchy_radio_camel.lvl4)",
        "unordered(hierarchy_radio.lvl4)",
        "unordered(hierarchy_radio_camel.lvl5)",
        "unordered(hierarchy_radio.lvl5)",
        "unordered(hierarchy_radio_camel.lvl6)",
        "unordered(hierarchy_radio.lvl6)",
        "unordered(hierarchy_camel.lvl0)",
        "unordered(hierarchy.lvl0)",
        "unordered(hierarchy_camel.lvl1)",
        "unordered(hierarchy.lvl1)",
        "unordered(hierarchy_camel.lvl2)",
        "unordered(hierarchy.lvl2)",
        "unordered(hierarchy_camel.lvl3)",
        "unordered(hierarchy.lvl3)",
        "unordered(hierarchy_camel.lvl4)",
        "unordered(hierarchy.lvl4)",
        "unordered(hierarchy_camel.lvl5)",
        "unordered(hierarchy.lvl5)",
        "unordered(hierarchy_camel.lvl6)",
        "unordered(hierarchy.lvl6)",
        "content",
      ],
      distinct: true,
      attributeForDistinct: "url",
      customRanking: [
        "desc(weight.pageRank)",
        "desc(weight.level)",
        "asc(weight.position)",
      ],
      ranking: [
        "words",
        "filters",
        "typo",
        "attribute",
        "proximity",
        "exact",
        "custom",
      ],
      highlightPreTag: '<span class="algolia-docsearch-suggestion--highlight">',
      highlightPostTag: "</span>",
      minWordSizefor1Typo: 3,
      minWordSizefor2Typos: 7,
      allowTyposOnNumericTokens: false,
      minProximity: 1,
      ignorePlurals: true,
      advancedSyntax: true,
      attributeCriteriaComputedByMinProximity: true,
      removeWordsIfNoResults: "allOptional",
    },
  },
});
```

**Reply by lslzl3000** [reply]
> Can you try if [#1935 (comment)](https://github.com/vuejs/vitepress/issues/1935#issuecomment-1433765099) works?

By adding `lang` in  `selectors` and `custom_settings` in top level, it's working, thanks for your help

But the `lang` in `facetFilters`  will be ignored, it will take the `lang` value in VitePress  anyway
```js
defineConfig({
    ...

    lang: 'en-US', // the requests use html lang anyway
    algolia:{
        searchParameters:{
            facetFilters: ["lang:zh"] // not working
        }
    },

    ...
}
```
is this a bug?  I think it should use local `facetFilters` first, if no lang provided in `facetFilters`, then go to html lang


One more question, the result is right，but the ui shows '#' as title/head
![Screenshot 2023-03-16 at 14 20 49](https://user-images.githubusercontent.com/3398981/225532401-be13c9c7-539c-4cbf-bc66-668a0833c166.png)

Is this configureable in JS/Vitepress side?  or something wrong with my crawler config?

**Reply by brc-dd** [reply]
> Is this configureable in JS/Vitepress side? or something wrong with my crawler config?

Something is wrong with your config. Try using the config I shared above (the long one).

> is this a bug? I think it should use local facetFilters first, if no lang provided in facetFilters, then go to html lang

No. It's expected. People aren't supposed to manually specify the `lang:` filter. Our i18n algorithm handles that. Why do you have a `lang: 'en-US'` there at first place? Shouldn't that be `zh` if your site is in Chinese?

**Reply by lslzl3000** [reply]
Thanks, I will try change crawler config

The `facetFilters:["lang:zh"]` just for test, Algolia doc said `lang` could be changed in local `facetFilters`

**Reply by brc-dd** [reply]
Yeah you can totally skip that thing. If you are using locales, vitepress will automatically adjust the config for you so that whatever language docs your user is viewing, searching will show content relevant to that only.

https://vitepress.dev/guide/i18n
https://vitepress.dev/reference/default-theme-search#i18n

**Reply by lslzl3000** [reply]
the long config you provided, it is new v3 DocSearch, so many things missing in legacy crawl(only few filed in v2), no idea how to match the v3 result, I will keep tring

**Reply by brc-dd** [reply]
Ah, ok. Have you tried changing the selectors to what's mentioned in #1935?

My guess is you've wrong lvl0 and lvl1.

**Reply by lslzl3000** [reply]
Yes, wrong about lvl0 & lvl1，finally got it work, thanks for your help

**Comment by baumannzone** [comment]
Algolia is working! I just got accepted to join the program for free and they sent me the appId and the apiKey and now everything works as expected.
